File: src/main/java/org/noxet/noxetserver/commands/misc/DoAs.java
package org.noxet.noxetserver.commands.misc;

import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabExecutor;
import org.bukkit.entity.Player;
import org.bukkit.help.HelpMap;
import org.bukkit.help.HelpTopic;
import org.noxet.noxetserver.NoxetServer;
import org.noxet.noxetserver.messaging.NoxetErrorMessage;
import org.noxet.noxetserver.messaging.NoxetMessage;
import org.noxet.noxetserver.messaging.NoxetSuccessMessage;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;

public class DoAs implements TabExecutor {
    @Override
    @SuppressWarnings("NullableProblems")
    public boolean onCommand(CommandSender commandSender, Command command, String s, String[] strings) {
        if(!commandSender.isOp()) {
            new NoxetErrorMessage(NoxetErrorMessage.ErrorType.PERMISSION, "You must be an operator to impersonate players.").send(commandSender);
            return true;
        }

        if(strings.length == 0) {
            new NoxetErrorMessage(NoxetErrorMessage.ErrorType.ARGUMENT, "Provide a player to run command as.").send(commandSender);
            return true;
        }

        Player doAsPlayer = NoxetServer.getPlugin().getServer().getPlayer(strings[0]);

        if(doAsPlayer == null) {
            new NoxetErrorMessage(NoxetErrorMessage.ErrorType.ARGUMENT, "That is not a valid player.").send(commandSender);
            return true;
        }

        List<String> commandArgs = new ArrayList<>(Arrays.asList(strings));
        commandArgs.remove(0); // Remove the player name argument.

        Command commandToRun = NoxetServer.getPlugin().getServer().getPluginCommand(commandArgs.get(0));

        if(command.equals(commandToRun)) {
            new NoxetErrorMessage(NoxetErrorMessage.ErrorType.COMMON, "Recursive doas not allowed!").send(commandSender);
            return true;
        }

        StringBuilder commandWithArgs = new StringBuilder();

        int i = 0;
        for(String arg : commandArgs)
            commandWithArgs.append(i++ != 0 ? " " : "").append(arg);

        new NoxetMessage("Performing command as \"" + doAsPlayer.getName() + "\": §o/" + commandWithArgs).send(commandSender);
        doAsPlayer.performCommand(commandWithArgs.toString());
        new NoxetSuccessMessage("Command was performed on " + doAsPlayer.getName() + ".").send(commandSender);

        return true;
    }

    @Override
    @SuppressWarnings("NullableProblems")
    public List<String> onTabComplete(CommandSender commandSender, Command command, String s, String[] strings) {
        if(!commandSender.isOp())
            return null;

        List<String> completions = new ArrayList<>();

        if(strings.length == 1) {
            Player player = NoxetServer.getPlugin().getServer().getPlayer(strings[0]);
            if(player == null)
                return null;
            completions.add(player.getName());
        } else if(strings.length == 2) {
            Player player = NoxetServer.getPlugin().getServer().getPlayer(strings[0]);
            if(player == null)
                return null;
            HelpMap commandMap = NoxetServer.getPlugin().getServer().getHelpMap();
            Collection<HelpTopic> commands = commandMap.getHelpTopics();

            for(HelpTopic eachCommand : commands)
                try{
                    if(eachCommand.getName().startsWith("/") && eachCommand.canSee(player))
                        completions.add(eachCommand.getName().substring(1));
                } catch(IllegalStateException ignored) {}
        } else if(strings.length > 2) {
            Player player = NoxetServer.getPlugin().getServer().getPlayer(strings[0]);
            if(player == null)
                return null;
            Command doAsCommand = NoxetServer.getPlugin().getServer().getPluginCommand(strings[1]);
            if(doAsCommand == null || !doAsCommand.testPermissionSilent(player) || doAsCommand.equals(command))
                return null;

            List<String> commandArgs = Arrays.asList(strings);
            commandArgs.remove(0); // Player name
            commandArgs.remove(1); // Command name

            return doAsCommand.tabComplete(player, doAsCommand.getName(), commandArgs.toArray(new String[0]), player.getLocation());
        }

        return completions;
    }
}

File: src/main/java/org/noxet/noxetserver/commands/admin/DoAs.scala
package org.noxet.noxetserver.commands.admin

import org.bukkit.command.{Command, CommandSender, TabExecutor}
import org.bukkit.entity.Player
import org.bukkit.help.{HelpMap, HelpTopic}
import org.noxet.noxetserver.NoxetServer
import org.noxet.noxetserver.commands.RegisteredCommand
import org.noxet.noxetserver.messaging.{ErrorMessage, Message, SuccessMessage}

import scala.jdk.CollectionConverters.*

object DoAs extends OperatorTabExecutor, RegisteredCommand("doas", this):
  override def onOperatorCommand(commandSender: CommandSender, command: Command, s: String, strings: Array[String]): Boolean =
    strings match
      case Array(playerName) =>
        Option(NoxetServer.getPlugin.getServer.getPlayer(playerName)) match
          case Some(player) =>
            val commandArgs = strings.drop(1) // Remove the player name argument.

            val commandToRun = NoxetServer.getPlugin.getServer.getPluginCommand(commandArgs(0))

            if command == commandToRun then
              ErrorMessage(ErrorMessage.ErrorType.Common, "Recursive DOAS not allowed!").send(commandSender)
              return true

            val commandWithArgs = commandArgs.mkString(" ")

            Message(s"Performing command as \"${doAsPlayer.getName}\": §o/$commandWithArgs").send(commandSender)
            doAsPlayer.performCommand(commandWithArgs)
            SuccessMessage(s"Command was performed on ${doAsPlayer.getName}.").send(commandSender)
            true
          case None =>
            ErrorMessage(ErrorMessage.ErrorType.Argument, "That is not a valid player.").send(commandSender)
            true
      case _ =>
        ErrorMessage(ErrorMessage.ErrorType.Argument, "Provide a player to run command as.").send(commandSender)
        true

  override def onOperatorTabComplete(commandSender: CommandSender, command: Command, s: String, strings: Array[String]): List[String] =
    strings match
      case Array(playerName, _*) =>
        Option(NoxetServer.getPlugin.getServer.getPlayer(playerName)) match
          case Some(player) =>
            strings match
              case Array(_) => List(player.getName)
              case Array(_, _) =>
                NoxetServer.getPlugin.getServer.getHelpMap.commandMap.getHelpTopics.asScala
                  .filter(c => c.getName.startsWith("/") && c.canSee(player))
                  .map(_.getName.substring(1))
              case Array(_, testCmd, args*) =>
                Option(NoxetServer.getPlugin.getServer.getPluginCommand(testCmd)) match
                  case Some(doAsCmd) if doAsCmd.testPermissionSilent(player) && doAsCmd != command =>
                    doAsCmd.tabComplete(
                      player,
                      doAsCommand.getName,
                      args,
                      player.getLocation).asScala
                  case None => List()
          case None => List()
      case _ => List()